File: Sources/Utility.m
#import "Utility.h"
#import "NSArray-Extension.h"
#import "TeX2img-Swift.h"

NSString* execCommand(NSString *cmdline)
{
    NSTask *task = [NSTask new];
    NSPipe *pipe = [NSPipe pipe];
    task.launchPath = BASH_PATH;
    task.arguments = @[@"-c", cmdline];
    task.standardOutput = pipe;
    [task launch];
    [task waitUntilExit];
    
    return pipe.stringValue;
}

NSString* getFullPath(NSString *aPath)
{
    NSURL *url = [NSURL fileURLWithPath:aPath];
    if (!url) return nil;
    
    return [NSString stringWithUTF8String:url.fileSystemRepresentation].stringByStandardizingPath;
}

void previewFiles(NSArray<NSString*> *files, NSString *app)
{
    if (@available(macOS 10.15, *)) {
        NSArray<NSURL*> *targetURLs = [files mapUsingBlock: ^NSURL*(NSString* path){ return [NSURL fileURLWithPath:path]; }];
        [[NSWorkspace sharedWorkspace] openURLs:targetURLs
                           withApplicationAtURL:[NSURL fileURLWithPath:app]
                                  configuration:[NSWorkspaceOpenConfiguration configuration]
                              completionHandler:nil];
    } else {
        [files enumerateObjectsUsingBlock:^(NSString * _Nonnull path, NSUInteger idx, BOOL * _Nonnull stop) {
            [[NSWorkspace sharedWorkspace] openFile:path withApplication:app];
        }];
    }

}

BOOL isTeX2imgAnnotation(PDFAnnotation *annotation)
{
    if (![annotation.type isEqualToString:@"Text"]) {
        return NO;
    }
    
    NSString *contents = [annotation.contents stringByReplacingOccurrencesOfString:@"\r\n" withString:@"\n"];
    
    return ([contents rangeOfString:AnnotationHeader].location == 0);
}

File: Sources/Utility/Utility.swift
import AppKit
import Foundation
import PDFKit

private let annotationHeader = "%%TeX2img Document\n"
class Utility: NSObject {
    static func execCommand(_ executablePath: String, arguments: [String] = []) -> String {
        let task = Process()
        let pipe = Pipe()
        task.executableURL = URL(fileURLWithPath: executablePath)
        task.arguments = arguments
        task.standardOutput = pipe
        try? task.run()
        task.waitUntilExit()
        return pipe.stringValue
    }

    static func getFullPath(_ aPath: String) -> String? {
        guard !aPath.isEmpty else { return nil }
        return URL(fileURLWithPath: aPath).standardizedFileURL.path
    }

    static func previewFiles(_ files: [String], app: String) {
        if #available(macOS 10.15, *) {
            let targetURLs = files.map { URL(fileURLWithPath: $0) }
            NSWorkspace.shared.open(targetURLs,
                                    withApplicationAt: URL(fileURLWithPath: app),
                                    configuration: NSWorkspace.OpenConfiguration(),
                                    completionHandler: nil)
        } else {
            for path in files {
                NSWorkspace.shared.openFile(path, withApplication: app)
            }
        }
    }

    static func isTeX2imgAnnotation(_ annotation: PDFAnnotation) -> Bool {
        guard annotation.type == "Text" else { return false }
        let contents = annotation.contents?.replacingOccurrences(of: "\r\n", with: "\n") ?? ""
        return contents.hasPrefix(annotationHeader)
    }
}